src/main/java/minijava/ir/optimize/JmpBlockRemover.java

```diff
@@ -26,33 +31,47 @@ public boolean optimize(Graph graph) {
 
   @Override
   public void visit(Block block) {
-    if (isJmpBlock(block) && hasOnePredecessor(block)) {
-      hasChanged = true;
-      remove(block);
-    }
+    tryToIdentifyNewEdge(block)
+        .ifPresent(
+            edge -> {
+              hasChanged = true;
+              remove(block, edge.v1, edge.v2, edge.v3);
+            });
   }
 
-  private boolean isJmpBlock(Block block) {
+  private Optional<Tuple3<Node, Integer, Block>> tryToIdentifyNewEdge(Block block) {
```

Comment:
I don't like this method signature very much. We can't stop you from thinking functional, can we? ;)
I'll let this pass, but next time we should really think about using Closure instead! :smile: